example/test.py

```diff
@@ -0,0 +1,598 @@
+#!/usr/bin/env python3
+import json
+import re
+import sys
+import _thread as thread
+import threading
+import unittest
+from datetime import datetime, timedelta, timezone
+
+import docker
+import iso8601
+import requests
+from testtools.assertions import assert_that
+from testtools.matchers import (
+    AfterPreprocessing as After, Contains, Equals, GreaterThan, HasLength, Is,
+    LessThan, MatchesAll, MatchesAny, MatchesDict, MatchesListwise,
+    MatchesRegex, MatchesSetwise, Not)
+
+POSTGRES_IMAGE = 'postgres:9.6-alpine'
+POSTGRES_PARAMS = {
+    'service': 'db',
+    'db': 'mysite',
+    'user': 'mysite',
+    'password': 'secret',
+}
+RABBITMQ_IMAGE = 'rabbitmq:3.6-alpine'
+RABBITMQ_PARAMS = {
+    'service': 'amqp',
+    'vhost': '/mysite',
+    'user': 'mysite',
+    'password': 'secret',
+}
+DJANGO_BOOTSTRAP_IMAGE = 'mysite:py3'
+DATABASE_URL = (
+    'postgres://{user}:{password}@{service}/{db}'.format(**POSTGRES_PARAMS))
+BROKER_URL = (
+    'amqp://{user}:{password}@{service}/{vhost}'.format(**RABBITMQ_PARAMS))
+
+
+def resource_name(name, namespace='test'):
+    return '{}_{}'.format(namespace, name)
+
+
+def quit_function(fn_name):
+    # https://stackoverflow.com/a/31667005
+    print('{} took too long'.format(fn_name), file=sys.stderr)
+    sys.stderr.flush()  # Python 3 stderr is likely buffered.
+    # FIXME: Interrupting the main thread is hacky
+    thread.interrupt_main()  # raises KeyboardInterrupt
+
+
+def exit_after(s):
+    """
+    Use as decorator to exit process if function takes longer than s seconds
+    https://stackoverflow.com/a/31667005
+    """
+    def outer(fn):
+        def inner(*args, **kwargs):
+            timer = threading.Timer(s, quit_function, args=[fn.__name__])
+            timer.start()
+            try:
+                result = fn(*args, **kwargs)
+            finally:
+                timer.cancel()
+            return result
+        return inner
+    return outer
+
+
+@exit_after(10)
+def wait_for_log_line(container, pattern):
+    for line in container.logs(stream=True):
+        line = line.decode('utf-8').rstrip()  # Drop the trailing newline
+        if re.search(pattern, line):
+            return line
+
+
+def output_lines(raw_output, encoding='utf-8'):
+    decoded = raw_output.decode(encoding)
+    lines = decoded.split('\n')
+    # Remove extra newline at end of output
+    lines.pop()
+    return lines
+
+
+def list_container_processes(container):
+    ps_output = container.exec_run(
+        ['ps', 'ax', '-o', 'pid,ruser,command', '--no-headers'])
+    ps_lines = output_lines(ps_output)
```

Comment:
Yeah well it's not quite that :/

It seems to basically be `ps` from _outside_ the container, so it returns the host PID and uses the host UID <-> username mapping. It also seems to use the `ps` binary on the host, meaning the options may be a bit different depending on platform. This is annoying but I guess you can't be guaranteed to even have `ps` inside the container so it makes sense in a way.

It does return stuff in a nice dict so it would be nice to use but it doesn't return stuff in the namespaces we want.